### PR TITLE
Unify VM and Instance classes

### PR DESCRIPTION
**Pull request by mfalesni**

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit fdf15dcb5c71b284c162dc69e913aef20e5e84a3:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/infrastructure/virtual_machines.py:7:1: F401 'js' imported but unused
./cfme/infrastructure/virtual_machines.py:9:1: F401 'TemplateNotFound' imported but unused
./cfme/infrastructure/virtual_machines.py:12:1: F401 'mixins' imported but unused
./cfme/infrastructure/virtual_machines.py:17:1: F401 'date' imported but unused
./cfme/infrastructure/virtual_machines.py:18:1: F401 'contextmanager' imported but unused
./cfme/infrastructure/virtual_machines.py:23:1: F401 'Pretty' imported but unused
./cfme/infrastructure/virtual_machines.py:24:1: F401 'parsetime' imported but unused
./cfme/infrastructure/virtual_machines.py:25:1: F401 'deploy_template' imported but unused
./cfme/infrastructure/virtual_machines.py:26:1: F401 'TimedOutError' imported but unused
./cfme/infrastructure/virtual_machines.py:27:1: F401 'ActionNotSupported' imported but unused
./cfme/infrastructure/virtual_machines.py:27:1: F401 'VMInstanceNotFound' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 87fa1c5c6d1702739452359903f88b3a314b3d35:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/infrastructure/virtual_machines.py:7:1: F401 'js' imported but unused
./cfme/infrastructure/virtual_machines.py:9:1: F401 'TemplateNotFound' imported but unused
./cfme/infrastructure/virtual_machines.py:12:1: F401 'mixins' imported but unused
./cfme/infrastructure/virtual_machines.py:17:1: F401 'date' imported but unused
./cfme/infrastructure/virtual_machines.py:18:1: F401 'contextmanager' imported but unused
./cfme/infrastructure/virtual_machines.py:23:1: F401 'Pretty' imported but unused
./cfme/infrastructure/virtual_machines.py:24:1: F401 'parsetime' imported but unused
./cfme/infrastructure/virtual_machines.py:25:1: F401 'deploy_template' imported but unused
./cfme/infrastructure/virtual_machines.py:26:1: F401 'TimedOutError' imported but unused
./cfme/infrastructure/virtual_machines.py:27:1: F401 'ActionNotSupported' imported but unused
./cfme/infrastructure/virtual_machines.py:27:1: F401 'VMInstanceNotFound' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 13552ad52ba829b238c99274dd38df937d13e16b:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/infrastructure/virtual_machines.py:7:1: F401 'js' imported but unused
./cfme/infrastructure/virtual_machines.py:9:1: F401 'TemplateNotFound' imported but unused
./cfme/infrastructure/virtual_machines.py:12:1: F401 'mixins' imported but unused
./cfme/infrastructure/virtual_machines.py:17:1: F401 'date' imported but unused
./cfme/infrastructure/virtual_machines.py:18:1: F401 'contextmanager' imported but unused
./cfme/infrastructure/virtual_machines.py:23:1: F401 'Pretty' imported but unused
./cfme/infrastructure/virtual_machines.py:24:1: F401 'parsetime' imported but unused
./cfme/infrastructure/virtual_machines.py:25:1: F401 'deploy_template' imported but unused
./cfme/infrastructure/virtual_machines.py:26:1: F401 'TimedOutError' imported but unused
./cfme/infrastructure/virtual_machines.py:27:1: F401 'ActionNotSupported' imported but unused
./cfme/infrastructure/virtual_machines.py:27:1: F401 'VMInstanceNotFound' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit e19709567c0d2682cc6834e568852d65a47712f4:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 860edf1101e9a27c0c441b6c191cf84cb725f086:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

File: cfme/infrastructure/virtual_machines.py [hunk 1]
```diff
@@ -101,106 +77,117 @@ def date_retire_element(fill_data):
...
+
+
+def extend_nav(cls):
+    nav.add_branch(cls.__name__, scavenge_class(cls, ignore_navigate=True))
+
+
+@extend_nav
+class infrastructure_virtual_machines(object):
+    class infra_vm_and_templates(object):
+        @staticmethod
```
Comment (psav) [inline cfme/infrastructure/virtual_machines.py]: Even though I'm still not sure about this redesign, we could lose some lines, by making a metaclass that wraps and replaces the navigate method with a staticmethod decorated version.

Comment (psav) [inline cfme/infrastructure/virtual_machines.py]: Maybe something like this? @seandst @mfalesni did I do a bad thing below?
```python
class AttrMetaclass(type):

    def __new__(cls, clsname, bases, dct):

        attrs = {}
        for name, val in dct.items():
            if name == "navigate":
                 attrs[name] = staticmethod(val)
            else:
                 attrs[name] = val
        return super(AttrMetaclass, cls).__new__(cls, clsname, bases, attrs)

class Intermediate(object):
    __metaclass__ = AttrMetaclass


class nav_item(Intermediate):
    def navigate():
        print "p"


l = nav_item()
l.navigate()
```

Comment (mfalesni) [inline cfme/infrastructure/virtual_machines.py]: Yup, I want to do something like that. Will check and if it works, I will incorporate it in the current proposal.

**Comment by mfalesni** [comment]
@psav I did it the other way, I just check if it is an unbound method (since I don't instantiate the class structure), I just read the im_func of the unbound method. There is also a benefit that when the functions are defined like this, it helps debugging.

**Comment by dajoRH** [comment]
Lint report for commit 9cd2ddd8d662630543d6b27eedc7544e101e1b33:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 6b103f58706bc6462f92227f3235da270aa50054:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 1e553bd9c0fd8bb0ba819244ce3e3af83420fe8a:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 553e3616b0a324a332a7d9315d1e9fa974601626:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 28d575db994dcce591c6cc20bacde0351967b17e:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit fc2636d9dad14de8db8a4bec0bccc1f97bed5ecd:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 325ce51e48305ce896302bfcadd31163d07bff38:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 6797726d9d3ab881ebb97ccef6c6d4a164302e49:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit e2fed441f570659f498805103d35081c5480484b:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a557cb01297817be2312c88a711a8d23d13a0583:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/cloud/instance.py:9:1: F401 'form_buttons' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 69b54022ec61a9434f50de0a74bac53405dd8c28:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 1b009fa29c15fdc21d5e1a8be9e6bd7f040fb2e3:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8772d89ba3058a1108aee78e51ace4296b64011a:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 4cfde7d6ae30ada432d934f0f0e725725f31fffd:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by mfalesni** [comment]
@psav @seandst Tests appear to be quite well-behaved so far, but that is just for clouds now. I will then have to finish the documentation and reorganize commits (will squash, but not in a single one probably). And also try to ack one last time if something old snuck in meanwhile.

**Comment by dajoRH** [comment]
Lint report for commit 9dfbe88741f8697f5a7381c5a6c7000e439c6a9c:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit cdfdcdddc6f46d8163797cb8b9c84d9ecba9dc1e:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 20379beed3ba5c956e68ed51e98315e547b17c82:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 0b80d7ae45f669efb7c2275d861c468bb8f841a5:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 1a009dd7e14a3efda69e011a49efb4b82aafee3f:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 96360f7c8cd77a088fac1d118d2cd83ae534d942:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8b2e2b3a9e6e70ae65ca863833f6bf95846201ff:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit f878625268aa766d3d17c248074ff081db8803d0:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 9278073c7ec18d9bcd68fe03312bbeb8f32175bc:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by seandst** [comment]
PRT results have been inconclusive, but what we've gotten back from it, as well as what we've seen in manual testing, looks promising. I think we'll just go ahead with the merge and deal with any aftermath as-needed.

**Pull request merged**